src/engine/renderer/particle_system.cpp

```diff
@@ -1,47 +1,51 @@
 namespace Renderer {
 
-bool Particle::dead() {
-    return progress > 1.0;
-}
-
 void Particle::update(f32 delta) {
-    if (dead()) return;
+    alive = alive && (keep_alive || progress < 1.0);
+    if (!alive) return;
     progress += inv_alive_time * delta;
     velocity += acceleration * delta;
     position += velocity * delta;
     velocity *= pow(damping, delta);
-
     rotation += angular_velocity * delta;
 }
 
 void Particle::render(u32 layer, Vec2 origin, s32 slot, Vec2 uv_min, Vec2 uv_dim) {
-    if (dead()) return;
+    if (!alive) return;
+    f32 progress_mod = MOD(progress, 1.0);
     Renderer::push_sprite_rect(
         layer,
         slot,
         position + origin,
-        dim * LERP(spawn_size, progress, die_size),
+        dim * (*progress_func_size)(first_size, first_size_deriv, second_size, second_size_deriv, progress_mod),
         rotation,
         uv_min,
         uv_dim,
-        LERP(spawn_color, progress, die_color));
+        (*progress_func_color)(first_color, first_color_deriv, second_color, second_color_deriv, progress_mod)
+    );
 }
 
-
 Particle ParticleSystem::generate() {
     ASSERT(particles, "Trying to use uninitalized/destroyed particle system");
 
     f32 first_size = spawn_size.random();
+    f32 first_size_deriv = spawn_size_deriv.random();
     f32 second_size = one_size ? first_size : die_size.random();
+    f32 second_size_deriv = die_size_deriv.random();
```

Comment:
This name is kinda confusing, since you might mean the second derivative, not the other derivative. So it can cause ambiguity. Same applies to the color.